### PR TITLE
Deprecate legacy code

### LINKED ISSUE #749: refactor: deprecate legacy code
### What is the expected enhancement?

#746 migrated (almost) the entire code base into a new structure.
While doing so, we did not yet add deprecation warnings for the legacy code. This issue can be closed, once all legacy modules have deprecation warnings associated with them.

Some caveats:
- the `BOPESSampler`-related code has not been migrated. However, it should still be deprecated. A feature request will be opened shortly explaining how we may achieve warm-starting variational algorithms in the future. This will be the de-facto replacement of the `BOPESSampler`.
- the protein folding code has also not been migrated. #742 reflects that this code will indeed be extracted into another repository in its entirety. Maybe we should hold off with the deprecation warnings on that particular code, until the extraction has been done.

### PR DESCRIPTION
**Pull request by manoelmarques**
### Summary
Closes #749

### CODE REVIEW

File: qiskit_nature/algorithms/pes_samplers/__init__.py
Comment (mrossinek) [inline qiskit_nature/algorithms/pes_samplers/__init__.py:79]: Should we refer to #750 to some degree here? It might result in at least the `extrapolators` going somewhere else..

File: qiskit_nature/drivers/bosonic_bases/__init__.py
Comment (mrossinek) [inline qiskit_nature/drivers/bosonic_bases/__init__.py:37]: Actually, these should all be already deprecated, since they were succeeded by the bases in the properties framework.
If they were not, I don't think its a big problem since the only place to use them was with the already deprecated `WatsonHamiltonian`

The reason that `qiskit_nature.second_q.drivers.bosonic_bases` exists, is because we left the removal of deprecated code to #748

File: qiskit_nature/drivers/qmolecule.py
Comment (mrossinek) [inline qiskit_nature/drivers/qmolecule.py:56]: This class will also be removed as part of #748. No need to update the deprecation here

File: qiskit_nature/drivers/watson_hamiltonian.py
Comment (mrossinek) [inline qiskit_nature/drivers/watson_hamiltonian.py:36]: Same holds as for `QMolecule`

File: qiskit_nature/properties/__init__.py
Comment (mrossinek) [inline qiskit_nature/properties/__init__.py:43]: This is not entirely true. Some of the properties will end up in `qiskit_nature.second_q.hamiltonians`.
Some other utilities in here (e.g. the `Integrals` classes) will be replaced by the `PolynomialTensor`.

I think this requires a bit more granular work unfortunately.

File: qiskit_nature/problems/second_quantization/__init__.py
Comment (mrossinek) [inline qiskit_nature/problems/second_quantization/__init__.py:47]: The `lattice` stack partially moved to `hamiltonians`

**Review by mrossinek — COMMENTED** [review]
Thanks for getting started on this, Manoel!
I'm afraid we may need to do a few of these a bit more granular..

File: qiskit_nature/properties/__init__.py
Comment (manoelmarques) [reply qiskit_nature/properties/__init__.py]: Where do the properties go ? Should we wait until classes like  `PolynomialTensor` are merged so that we add proper deprecation messages ?

File: qiskit_nature/problems/second_quantization/__init__.py
Comment (manoelmarques) [reply qiskit_nature/problems/second_quantization/__init__.py]: I changed the deprecation msg location but should we wait until the stack is moved to add the deprecation msgs. for this ?

File: qiskit_nature/properties/__init__.py
Comment (mrossinek) [reply qiskit_nature/properties/__init__.py]: Here is an overview of the classes in properties and where they will end up:
```
GroupedProperty -> removed, no replacement necessary
Property -> refactored, base class as this no longer existed in the same form (if at all, it will be `qiskit_nature.second_q.properties.Property`)

DriverMetadata -> qiskit_nature.second_q.problems.DriverMetadata (maybe rename to `Precedence` using naming from QCSchema)

AngularMomentum -> qiskit_nature.second_q.properties
Magnetization -> qiskit_nature.second_q.properties
ParticleNumber -> qiskit_nature.second_q.properties
(Electronic)DipoleMoment -> qiskit_nature.second_q.properties
ElectronicEnergy -> qiskit_nature.second_q.hamiltonians
ElectronicIntegrals -> succeeded by qiskit_nature.second_q.operators.PolynomialTensor
OneBodyElectronicIntegrals -> PolynomialTensor
TwoBodyElectronicIntegrals -> PolynomialTensor
IntegralProperty -> refactored, no longer needed as such
ElectronicStructureDriverResult -> succeeded by ElectronicStructureProblem

OccupiedModals -> qiskit_nature.second_q.properties
VibrationalEnergy -> qiskit_nature.second_q.hamiltonians
VibrationalIntegrals -> succeeded by qiskit_nature.second_q.operators.PolynomialTensor
VibrationalStructureDriverResult -> succeeded by VibrationalStructureProblem

ElectronicBasis -> TBD
VibrationalBasis -> TBD
```

File: qiskit_nature/algorithms/ground_state_solvers/adapt_vqe.py
Comment (mrossinek) [inline qiskit_nature/algorithms/ground_state_solvers/adapt_vqe.py:98]:
```suggestion
            new_name="qiskit.algorithms.minimum_eigen_solvers.AdaptVQE",
```

File: qiskit_nature/problems/second_quantization/lattice/__init__.py
Comment (mrossinek) [inline qiskit_nature/problems/second_quantization/lattice/__init__.py:66]: I think we need to differentiate here:
- the `LatticeModelProblem` class did indeed go to `second_q.problems`
- but the `lattice` submodule will go to `second_q.hamiltonians`

File: qiskit_nature/properties/grouped_property.py
Comment (mrossinek) [inline qiskit_nature/properties/grouped_property.py:58]: This class is actually going away in its entirety in favor of storing the properties directly on the `BaseProblem` objects.
Maybe we can refer to #705 for more information?

File: qiskit_nature/properties/second_quantization/driver_metadata.py
Comment (mrossinek) [inline qiskit_nature/properties/second_quantization/driver_metadata.py:46]: This class will likely also be removed in its entirety.
A replacement is not yet planned as this information is not relevant to the stack itself, thus, is not going to be propagated down in the new workflow (until it becomes necessary per a specific feature request). Again, we could link to #705

File: qiskit_nature/properties/second_quantization/electronic/bases/__init__.py
Comment (mrossinek) [inline qiskit_nature/properties/second_quantization/electronic/bases/__init__.py:33]:
```suggestion
    old_name="qiskit_nature.properties.second_quantization.electronic.bases",
```

File: qiskit_nature/properties/second_quantization/electronic/bases/__init__.py
Comment (mrossinek) [inline qiskit_nature/properties/second_quantization/electronic/bases/__init__.py:35]: After the latest design plans (yet to be published here on github), this will actually go here instead:
```suggestion
    new_name="qiskit_nature.second_q.problems.bases",
```

File: qiskit_nature/properties/second_quantization/electronic/dipole_moment.py
Comment (mrossinek) [inline qiskit_nature/properties/second_quantization/electronic/dipole_moment.py:71]: This file actually contains another class: `ElectronicDipoleMoment` which also needs to be deprecated.
I might improve upon this more later, but I will update the deprecation messages accordingly when it comes to that :+1:

File: qiskit_nature/properties/second_quantization/vibrational/bases/__init__.py
Comment (mrossinek) [inline qiskit_nature/properties/second_quantization/vibrational/bases/__init__.py:33]:
```suggestion
    old_name="qiskit_nature.properties.second_quantization.vibrational.bases",
```

File: qiskit_nature/properties/second_quantization/vibrational/bases/__init__.py
Comment (mrossinek) [inline qiskit_nature/properties/second_quantization/vibrational/bases/__init__.py:35]:
```suggestion
    new_name="qiskit_nature.second_q.problems.bases",
```

File: qiskit_nature/properties/second_quantization/vibrational/vibrational_energy.py
Comment (mrossinek) [inline qiskit_nature/properties/second_quantization/vibrational/vibrational_energy.py:73]:
```suggestion
            new_type=DeprecatedType.CLASS,
            new_name="qiskit_nature.second_q.hamiltonians.ElectronicEnergy",
```

**Review by mrossinek — COMMENTED** [review]
Some more details, but other than this I believe this looks good :+1:

File: qiskit_nature/properties/second_quantization/vibrational/vibrational_energy.py
Comment (mrossinek) [inline qiskit_nature/properties/second_quantization/vibrational/vibrational_energy.py:73]:
```suggestion
            new_name="qiskit_nature.second_q.hamiltonians.VibrationalEnergy",
```

**Review by mrossinek — COMMENTED** [review]
I think there are still deprecations missing on:
- `qiskit_nature/converters -> qiskit_nature/second_q/mappers`
- `qiskit_nature/results -> qiskit_nature/second_q/problems`

Deprecations for `qiskit_nature/runtime` can be added once we sorted out the algorithms situation.

Other than that, only a release note should be added

**Review by mrossinek — APPROVED** [review]
LGTM now, thanks Manoel!

Let's give @woodsp-ibm a chance to review this before merging though :+1:

File: releasenotes/notes/deprecate-legacy-code-fe9db8a009b6f5d2.yaml
Comment (woodsp-ibm) [inline releasenotes/notes/deprecate-legacy-code-fe9db8a009b6f5d2.yaml:1]: Maybe this PR is not the place, but at some point I think a prelude section that outlines the new direction for the end user reading the release note and talks a bit about the deprecation and migration path. Maybe we can wait until we have a migration guide ready and have this aspect of the release note be part of that. This then may help a reader to understand the implications to what they currently have. I find the deprecation note as it stands here does not really help me too much - it says "all code that was migrated", so does this impact me or not reading this would be my question. (The answer is pretty much always going to be yes since pretty much all code was migrated!) With this text as-is, to try to figure anything, I have to go to 749 which references another PR (722) which lists a bunch too. Perhaps with an overall description of the direction/migration it will be clear to user and this will end up being a bit of detail that this unless you trace through the ref'd PR and try to work things out yourself it will not offer much help! Maybe we could alter this to something like the following which is a bit more descriptive. (And maybe later we can even update this release note to refer the user back to the prelude we add)

The vast majority of the existing Nature code has been migrated/refactored to `qiskit_nature/second_q` and the corresponding existing code has all been deprecated according to issue:....

Comment (mrossinek) [reply releasenotes/notes/deprecate-legacy-code-fe9db8a009b6f5d2.yaml:1]: I absolutely agree! I was thinking that I could add one major release note at the end of the refactoring process, which explains the new workflow.

That being said, the contents of this particular release note could be improved. Maybe we can add a short overview of what has been deprecated in favor of where it goes? Maybe even in table format?
Maybe something with the following content in mind:
```
qiskit_nature.algorithms -> qiskit_nature.second_q.algorithms
qiskit_nature.algorithms.pes_samplers -> removed; refer to X for more details
qiskit_nature.circuit -> qiskit_nature.second_q.circuit
qiskit_nature.converters -> qiskit_nature.second_q.mappers
qiskit_nature.drivers -> qiskit_nature.second_q.drivers
qiskit_nature.mappers -> qiskit_nature.second_q.mappers
qiskit_nature.operators -> qiskit_nature.second_q.operators
qiskit_nature.problems -> qiskit_nature.second_q.problems
qiskit_nature.problems.sampling -> moved to prototypes; refer to X for more details
qiskit_nature.properties -> qiskit_nature.second_q.hamiltonians, qiskit_nature.second_q.properties (refer to X for more details)
qiskit_nature.results -> qiskit_nature.second_q.problems
qiskit_nature.transformers -> qiskit_nature.second_q.transformers
```

File: releasenotes/notes/deprecate-legacy-code-fe9db8a009b6f5d2.yaml
Comment (mrossinek) [inline releasenotes/notes/deprecate-legacy-code-fe9db8a009b6f5d2.yaml:8]: I'm not sure what formatting you attempted here. The output right now is as such:
screenshot_1659627292

The table looks nice imo, but maybe we can simply make this text as following:
```suggestion
    The following table provides a rough overview of the old and new package locations.
    In certain cases, single classes might have been treated differently.
    Be sure to check out the referenced issues or the migration guide for more details.
```

Where `"the migration guide"` will be added when the refactoring comes to and end, and will be linked to in its release note. We can then even go back into this reno, and make this a link, too :+1:

**Review by mrossinek — APPROVED** [review]
Thank you, Manoel!

Let's give @woodsp-ibm a chance to review this, too, before merging :+1:

File: releasenotes/notes/deprecate-legacy-code-fe9db8a009b6f5d2.yaml
Comment (woodsp-ibm) [inline releasenotes/notes/deprecate-legacy-code-fe9db8a009b6f5d2.yaml:9]: Juts a comment - when we add the migration guide perhaps we can remember to go back and edit the reference here to make it a link. But it all LGTM.

**Review by woodsp-ibm — APPROVED** [review]

**Pull request merged**
